File: internal/generator/sdkgen/test_unwrap_envelope.js
import { describe, expect, test } from "bun:test";
import { pathToFileURL } from "node:url";
import path from "node:path";

const sdkDir = process.env.SDK_DIR;
if (!sdkDir) {
  throw new Error("SDK_DIR env var required");
}

const { HttpClient } = await import(
  pathToFileURL(path.join(sdkDir, "client.ts")).href
);
const { NetworkError } = await import(
  pathToFileURL(path.join(sdkDir, "types.ts")).href
);

const ERR_EXPECT_OBJECT =
  "SDK contract error: expected RFC 9457 envelope object with data/meta.requestId in success response";
const ERR_EXPECT_FIELDS =
  "SDK contract error: expected RFC 9457 envelope fields data and meta in success response";
const ERR_EXPECT_REQUEST_ID =
  "SDK contract error: expected RFC 9457 envelope meta.requestId in success response";

function responseForPayload(payload) {
  if (payload === undefined) {
    return new Response("", { status: 200, statusText: "OK" });
  }
  return new Response(JSON.stringify(payload), {
    status: 200,
    statusText: "OK",
    headers: { "Content-Type": "application/json" },
  });
}

async function unwrapViaClient(payload) {
  const client = new HttpClient({
    baseUrl: "https://api.example.com",
    fetch: async () => responseForPayload(payload),
  });
  return client.get("/v1/x");
}

describe("strict envelope unwrap via generated HttpClient", () => {
  const successCases = [
    {
      name: "valid envelope object",
      payload: { data: { id: "x" }, meta: { requestId: "abc" } },
      expected: { id: "x" },
    },
    {
      name: "valid envelope null data",
      payload: { data: null, meta: { requestId: "abc" } },
      expected: null,
    },
    {
      name: "valid envelope array data",
      payload: { data: [1, 2, 3], meta: { requestId: "abc" } },
      expected: [1, 2, 3],
    },
    {
      name: "valid envelope string data",
      payload: { data: "hello", meta: { requestId: "abc" } },
      expected: "hello",
    },
    {
      name: "extra keys are allowed",
      payload: { data: { id: "x" }, meta: { requestId: "abc" }, links: {} },
      expected: { id: "x" },
    },
    {
      name: "empty requestId is accepted",
      payload: { data: { id: "x" }, meta: { requestId: "" } },
      expected: { id: "x" },
    },
    {
      name: "undefined payload is passthrough",
      payload: undefined,
      expected: undefined,
    },
  ];

  for (const tc of successCases) {
    test(tc.name, async () => {
      await expect(unwrapViaClient(tc.payload)).resolves.toEqual(tc.expected);
    });
  }

  const errorCases = [
    {
      name: "null payload is contract violation",
      payload: null,
      expectedError: ERR_EXPECT_OBJECT,
    },
    {
      name: "string payload is contract violation",
      payload: "hello",
      expectedError: ERR_EXPECT_OBJECT,
    },
    {
      name: "array payload is contract violation",
      payload: [1, 2, 3],
      expectedError: ERR_EXPECT_FIELDS,
    },
    {
      name: "no data key",
      payload: { result: { id: "x" }, meta: { requestId: "abc" } },
      expectedError: ERR_EXPECT_FIELDS,
    },
    {
      name: "no meta key",
      payload: { data: { id: "x" } },
      expectedError: ERR_EXPECT_FIELDS,
    },
    {
      name: "meta missing requestId",
      payload: { data: { id: "x" }, meta: { other: "v" } },
      expectedError: ERR_EXPECT_REQUEST_ID,
    },
    {
      name: "meta not object",
      payload: { data: { id: "x" }, meta: "bad" },
      expectedError: ERR_EXPECT_REQUEST_ID,
    },
  ];

  for (const tc of errorCases) {
    test(tc.name, async () => {
      try {
        await unwrapViaClient(tc.payload);
        throw new Error("expected unwrapViaClient to throw");
      } catch (error) {
        expect(error).toBeInstanceOf(NetworkError);
        expect(error.message).toBe("Request error: " + tc.expectedError);
      }
    });
  }
});

File: internal/generator/rustsdkgen/test_unwrap_envelope.rs
use std::collections::BTreeMap;

const ERR_EXPECT_OBJECT: &str =
    "SDK contract error: expected RFC 9457 envelope object with data/meta.requestId in success response";
const ERR_EXPECT_FIELDS: &str =
    "SDK contract error: expected RFC 9457 envelope fields data and meta in success response";
const ERR_EXPECT_REQUEST_ID: &str =
    "SDK contract error: expected RFC 9457 envelope meta.requestId in success response";

#[derive(Clone, Debug, PartialEq)]
enum Value {
    Object(BTreeMap<String, Value>),
    Array(Vec<Value>),
    String(String),
    Number(i64),
    Bool(bool),
    Null,
}

fn object(entries: Vec<(&str, Value)>) -> Value {
    let mut map = BTreeMap::new();
    for (key, value) in entries {
        map.insert(key.to_string(), value);
    }
    Value::Object(map)
}

fn unwrap_envelope(payload: Value) -> Result<Value, String> {
    let object = match payload {
        Value::Object(map) => map,
        _ => return Err(ERR_EXPECT_OBJECT.to_string()),
    };

    let data = match object.get("data") {
        Some(value) => value.clone(),
        None => return Err(ERR_EXPECT_FIELDS.to_string()),
    };

    let meta = match object.get("meta") {
        Some(value) => value,
        None => return Err(ERR_EXPECT_FIELDS.to_string()),
    };

    let meta_object = match meta {
        Value::Object(map) => map,
        _ => return Err(ERR_EXPECT_REQUEST_ID.to_string()),
    };

    if !meta_object.contains_key("requestId") {
        return Err(ERR_EXPECT_REQUEST_ID.to_string());
    }

    Ok(data)
}

#[test]
fn unwrap_envelope_success_cases() {
    let success_cases: Vec<(&str, Value, Value)> = vec![
        (
            "valid envelope object",
            object(vec![
                ("data", object(vec![("id", Value::String("x".to_string()))])),
                (
                    "meta",
                    object(vec![("requestId", Value::String("abc".to_string()))]),
                ),
            ]),
            object(vec![("id", Value::String("x".to_string()))]),
        ),
        (
            "valid envelope null data",
            object(vec![
                ("data", Value::Null),
                (
                    "meta",
                    object(vec![("requestId", Value::String("abc".to_string()))]),
                ),
            ]),
            Value::Null,
        ),
        (
            "valid envelope array data",
            object(vec![
                ("data", Value::Array(vec![Value::Number(1), Value::Number(2)])),
                (
                    "meta",
                    object(vec![("requestId", Value::String("abc".to_string()))]),
                ),
            ]),
            Value::Array(vec![Value::Number(1), Value::Number(2)]),
        ),
        (
            "empty requestId still accepted",
            object(vec![
                ("data", Value::Bool(true)),
                (
                    "meta",
                    object(vec![("requestId", Value::String("".to_string()))]),
                ),
            ]),
            Value::Bool(true),
        ),
    ];

    for (name, payload, expected) in success_cases {
        let actual = unwrap_envelope(payload).expect(name);
        assert_eq!(actual, expected, "{}", name);
    }
}

#[test]
fn unwrap_envelope_error_cases() {
    let error_cases: Vec<(&str, Value, &str)> = vec![
        ("no data key", object(vec![("meta", object(vec![("requestId", Value::String("abc".to_string()))]))]), ERR_EXPECT_FIELDS),
        ("no meta key", object(vec![("data", Value::String("x".to_string()))]), ERR_EXPECT_FIELDS),
        (
            "meta missing requestId",
            object(vec![
                ("data", Value::String("x".to_string())),
                ("meta", object(vec![("other", Value::String("v".to_string()))])),
            ]),
            ERR_EXPECT_REQUEST_ID,
        ),
        (
            "meta not object",
            object(vec![
                ("data", Value::String("x".to_string())),
                ("meta", Value::String("bad".to_string())),
            ]),
            ERR_EXPECT_REQUEST_ID,
        ),
        ("null payload", Value::Null, ERR_EXPECT_OBJECT),
        ("array payload", Value::Array(vec![]), ERR_EXPECT_OBJECT),
        ("number payload", Value::Number(42), ERR_EXPECT_OBJECT),
        ("bool payload", Value::Bool(false), ERR_EXPECT_OBJECT),
    ];

    for (name, payload, expected_error) in error_cases {
        let err = unwrap_envelope(payload).expect_err(name);
        assert_eq!(err, expected_error, "{}", name);
    }
}

#[test]
fn unwrap_envelope_returns_owned_data() {
    let baseline_data = object(vec![("id", Value::String("x".to_string()))]);
    let payload = object(vec![
        ("data", baseline_data.clone()),
        (
            "meta",
            object(vec![("requestId", Value::String("abc".to_string()))]),
        ),
    ]);

    let mut unwrapped = unwrap_envelope(payload).expect("unwrap should succeed");
    if let Value::Object(ref mut data_object) = unwrapped {
        data_object.insert("changed".to_string(), Value::Bool(true));
    } else {
        panic!("expected object payload");
    }

    assert_eq!(
        baseline_data,
        object(vec![("id", Value::String("x".to_string()))]),
    );
}
